File: service/api/src/main/java/org/delivery/api/domain/userorder/business/UserOrderBusiness.java
package org.delivery.api.domain.userorder.business;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.delivery.common.annotation.Business;
import org.delivery.api.domain.store.converter.StoreConverter;
import org.delivery.api.domain.store.service.StoreService;
import org.delivery.api.domain.storemenu.converter.StoreMenuConverter;
import org.delivery.api.domain.storemenu.service.StoreMenuService;
import org.delivery.api.domain.user.model.User;
import org.delivery.api.domain.userorder.controller.model.UserOrderDetailResponse;
import org.delivery.api.domain.userorder.controller.model.UserOrderRequest;
import org.delivery.api.domain.userorder.controller.model.UserOrderResponse;
import org.delivery.api.domain.userorder.converter.UserOrderConverter;
import org.delivery.api.domain.userorder.producer.UserOrderProducer;
import org.delivery.api.domain.userorder.service.UserOrderService;
import org.delivery.api.domain.userordermenu.converter.UserOrderMenuConverter;
import org.delivery.api.domain.userordermenu.service.UserOrderMenuService;
import org.delivery.db.store.StoreEntity;
import org.delivery.db.userordermenu.UserOrderMenuEntity;
import org.delivery.db.userordermenu.enums.UserOrderMenuState;

import java.util.List;
import java.util.stream.Collectors;

/*
@Business
@RequiredArgsConstructor
@Slf4j
public class UserOrderBusiness {
    private final UserOrderService userOrderService;
    private final StoreMenuService storeMenuService;
    private final UserOrderMenuService userOrderMenuService;
    private final StoreService storeService;


    private final UserOrderConverter userOrderConverter;
    private final UserOrderMenuConverter userOrderMenuConverter;
    private final StoreMenuConverter storeMenuConverter;
    private final StoreConverter storeConverter;

    private final UserOrderProducer userOrderProducer;
    private final ObjectMapper objectMapper;


    // 1. 사용자, 메뉴 id
    // 2. userOrder 생성
    // 3. userOrderMenu 생성
    // 4. 응답 생성
    public UserOrderResponse userOrder(
            @Valid
            User user,
            UserOrderRequest body
    ) {
        var storeEntity = storeService.getStoreWithThrow(body.getStoreId());

        var storeMenuEntityList = body.getStoreMenuIdList()
                .stream()
                .map(storeMenuService::getStoreMenuWithThrow)
                .toList();

        var userOrderEntity = userOrderConverter.toEntity(
                user,
                storeEntity,
                storeMenuEntityList);

        //주문
        var newUserOrderEntity = userOrderService.order(userOrderEntity);

        // 맵핑(user_order_menu)
        var userOrderMenuEntityList = storeMenuEntityList.stream()
                .map(it->{
                   // menu + user order
                   var userOrderMenuEntity = userOrderMenuConverter.toEntity(newUserOrderEntity, it);
                   return userOrderMenuEntity;
                }).toList();

        // 주문 내역에 기록 남기기
        userOrderMenuEntityList.forEach(userOrderMenuService::order);

        // 비동기로 가맹점에 주문 알리기
        userOrderProducer.sendOrder(newUserOrderEntity);

        // response
        return  userOrderConverter.toResponse(newUserOrderEntity);
    }


    public List<UserOrderDetailResponse> current(User user) {
        // 현재 사용자가 주문한 내역
        var userOrderEntityList = userOrderService.current(user.getId());

        var userOrderdetailResponseList = userOrderEntityList.stream()
                // 주문 1건씩 처리
                .map(userOrderEntity->{
                    log.info("사용자의 주문 정보 : {}", userOrderEntity);
          */
/*          try {
                        var jsonString = objectMapper.writeValueAsString(userOrderEntity);
                        log.info("Json String : {}", jsonString);

                    }catch (Exception e){
                        log.error("",e);
                    }*//*


                    // 사용자가 주문한 메뉴 리스트
                    // 기존 : var userOrderMenuEntityList = userOrderMenuService.getUserOrderMenu(it.getId());
                    // 엔티티 연관관계 설정 후 리펙토링
                    var userOrderMenuEntityList = userOrderEntity.getUserOrderMenuList().stream()
                            .filter(it-> it.getStatus().equals(UserOrderMenuState.REGISTERED))
                            .toList();

                    // 어떠한 메뉴를 주문했는지

                    */
/* var storeMenuEntityList = userOrderMenuEntityList.stream()
                            .map(userOrdermenuEntity->{
                                return storeMenuService.getStoreMenuWithThrow(userOrdermenuEntity.getStoreMenu().getId());
                            }).toList();*//*


                    // 연관관계 설정 후 리펙토링
                    var storeMenuEntityList = userOrderMenuEntityList.stream()
                            .map(UserOrderMenuEntity::getStoreMenu).toList();

                    // 사용자가 주문한 가게
                    // var storeEntity = storeService.getStoreWithThrow(storeMenuEntityList.stream().findFirst().get().getStore().getId());
                    // 연관관계 설정 후 리펙토링
                    var storeEntity = userOrderEntity.getStore();

                    return UserOrderDetailResponse.builder()
                            .userOrderResponse(userOrderConverter.toResponse(userOrderEntity))
                            .storeMenuResponseList(storeMenuConverter.toResponse(storeMenuEntityList))
                            .storeResponse(storeConverter.toResponse(storeEntity))
                            .build();

                }).toList();


        return userOrderdetailResponseList;
    }

    public List<UserOrderDetailResponse> history(User user) {
        // 현재 사용자가 주문한 내역
        var userOrderEntityList = userOrderService.history(user.getId());

        var userOrderdetailResponseList = userOrderEntityList.stream()
                // 주문 1건씩 처리
                .map(userOrderEntity->{

                    // 사용자가 주문한 메뉴 리스트
                    var userOrderMenuEntityList = userOrderEntity.getUserOrderMenuList();

                    // 어떠한 메뉴를 주문했는지
                    var storeMenuEntityList = userOrderMenuEntityList.stream()
                            .map(UserOrderMenuEntity::getStoreMenu).toList();

                    // 사용자가 주문한 가게
                    var storeEntity =  userOrderEntity.getStore();
                    return UserOrderDetailResponse.builder()
                            .userOrderResponse(userOrderConverter.toResponse(userOrderEntity))
                            .storeMenuResponseList(storeMenuConverter.toResponse(storeMenuEntityList))
                            .storeResponse(storeConverter.toResponse(storeEntity))
                            .build();

                }).toList();


        return userOrderdetailResponseList;
    }

    // 사용자 주문 메뉴 상세
    public UserOrderDetailResponse read(User user, Long orderId) {
        var userOrderEntity = userOrderService.getUserOrderWithOutStatusWithThrow(orderId,user.getId());

        var userOrderMenuEntityList = userOrderEntity.getUserOrderMenuList().stream()
                .filter(it -> it.getStatus().equals(UserOrderMenuState.REGISTERED))
                .toList();

        // 어떠한 메뉴를 주문했는지
        var storeMenuEntityList = userOrderMenuEntityList.stream()
                .map(UserOrderMenuEntity::getStoreMenu).toList();

        // 사용자가 주문한 가게
        var storeEntity = userOrderEntity.getStore();

        return UserOrderDetailResponse.builder()
                .userOrderResponse(userOrderConverter.toResponse(userOrderEntity))
                .storeMenuResponseList(storeMenuConverter.toResponse(storeMenuEntityList))
                .storeResponse(storeConverter.toResponse(storeEntity))
                .build();
    }
}
*/

File: service/api/src/main/kotlin/org/delivery/api/domain/userorder/business/UserOrderBusiness.kt
package org.delivery.api.domain.userorder.business

import org.delivery.api.common.Log
import org.delivery.api.domain.store.converter.StoreConverter
import org.delivery.api.domain.store.service.StoreService
import org.delivery.api.domain.storemenu.converter.StoreMenuConverter
import org.delivery.api.domain.storemenu.service.StoreMenuService
import org.delivery.api.domain.user.model.User
import org.delivery.api.domain.userorder.controller.model.UserOrderDetailResponse
import org.delivery.api.domain.userorder.controller.model.UserOrderRequest
import org.delivery.api.domain.userorder.controller.model.UserOrderResponse
import org.delivery.api.domain.userorder.converter.UserOrderConverter
import org.delivery.api.domain.userorder.producer.UserOrderProducer
import org.delivery.api.domain.userorder.service.UserOrderService
import org.delivery.api.domain.userordermenu.converter.UserOrderMenuConverter
import org.delivery.api.domain.userordermenu.service.UserOrderMenuService
import org.delivery.common.annotation.Business
import org.delivery.db.userordermenu.enums.UserOrderMenuState

@Business
class UserOrderBusiness(
    private val userOrderService: UserOrderService,
    private val userOrderConverter: UserOrderConverter,

    private val storeMenuService: StoreMenuService,
    private val storeMenuConverter: StoreMenuConverter,

    private val userOrderMenuService: UserOrderMenuService,
    private val userOrderMenuConverter: UserOrderMenuConverter,

    private val storeService: StoreService,
    private val storeConverter: StoreConverter,

    private val userOrderProducer: UserOrderProducer

) {
    companion object : Log

    fun userOrder(
        user: User?,
        body: UserOrderRequest?
    ): UserOrderResponse {
        // 가게 찾기
        val storeEntity = storeService.getStoreWithThrow(body?.storeId)

        // 주문한 메뉴 찾기
        val storeMenuEntityList = body?.storeMenuIdList
            ?.mapNotNull { it -> storeMenuService.getStoreMenuWithThrow(it) }
            ?.toList()

        // 주문하고나서의 반환값
        val userOrderEntity = userOrderConverter.toEntity(
            user = user,
            storeEntity = storeEntity,
            storeMenuEntityList = storeMenuEntityList
        ).run {
            //주문
            userOrderService.order(this)
        }

        // 맵핑[user order menu list 생성]
        val userOrderMenuEntityList = storeMenuEntityList
            ?.map { it -> userOrderMenuConverter.toEntity(userOrderEntity, it) }
            ?.toList()

        // 주문내역 기록 남기기
        userOrderMenuEntityList?.forEach { userOrderMenuService.order(it) }

        // 비동기 주문 알리기
        userOrderProducer.sendOrder(userOrderEntity)

        return userOrderConverter.toResponse(userOrderEntity)
    }

    fun current(
        user: User?
    ): List<UserOrderDetailResponse>? {
        return userOrderService.current(user?.id).map { userOrderEntity ->
            /*  val userOrderMenuEntityList = it.userOrderMenuList
                ?.filter {  it.status == UserOrderMenuState.REGISTERED }
                ?.toList()

                val storeMenuEntityList = userOrderMenuEntityList
                ?.map {it.storeMenu}
                ?.toList()
           */

            val storeMenuEntityList = userOrderEntity.userOrderMenuList?.stream()
                ?.filter { userOrderMenuEntity -> userOrderMenuEntity.status === UserOrderMenuState.REGISTERED }
                ?.map {
                    it.storeMenu
                }?.toList()


            val store = userOrderEntity.store
            UserOrderDetailResponse(
                userOrderResponse = userOrderConverter.toResponse(userOrderEntity),
                storeResponse = storeConverter.toResponse(store),
                storeMenuResponseList = storeMenuConverter.toResponse(storeMenuEntityList)
            )
        }.toList()
    }

    fun history(
        user: User?
    ): List<UserOrderDetailResponse>? {
        return userOrderService.history(user?.id).map { userOrderEntity ->
            val storeMenuEntityList = userOrderEntity.userOrderMenuList?.stream()
                ?.filter { userOrderMenuEntity -> userOrderMenuEntity.status === UserOrderMenuState.REGISTERED }
                ?.map {
                    it.storeMenu
                }?.toList()


            val store = userOrderEntity.store
            UserOrderDetailResponse(
                userOrderResponse = userOrderConverter.toResponse(userOrderEntity),
                storeResponse = storeConverter.toResponse(store),
                storeMenuResponseList = storeMenuConverter.toResponse(storeMenuEntityList)
            )
        }.toList()
    }

    fun read(
        user: User?,
        orderId: Long?

    ): UserOrderDetailResponse {
        return userOrderService.getUserOrderWithOutStatusWithThrow(orderId, user?.id)
            .let { userOrderEntity ->
                val storeMenuEntityList = userOrderEntity.userOrderMenuList
                    ?.stream()
                    ?.filter { it.status == UserOrderMenuState.REGISTERED }
                    ?.map{it.storeMenu}
                    ?.toList() // 있거나
                    ?:listOf()  // size 0

                UserOrderDetailResponse(
                    userOrderResponse = userOrderConverter.toResponse(userOrderEntity),
                    storeResponse = storeConverter.toResponse(userOrderEntity.store),
                    storeMenuResponseList = storeMenuConverter.toResponse(storeMenuEntityList)
                )

            }

    }

}